Major storage documentation overhaul (#363)
nais/doc: docs/persistence/README.md

Choose wisely.

```mermaid
graph TD
  A[I got data!] --> B[Is it structured?]
  B --> |Yes| C[Is it events?]
  B --> |No| D[Is it files?]
  C --> |Yes| Kafka
  C --> |No| E[Is it analytical?]
  D --> |Yes| GCS[Cloud Storage]
  D --> |No| Opensearch[OpenSearch]
  E --> |Yes| GBQ[BigQuery]
  E --> |No| GCSQL[Cloud SQL]

  click GBQ "#bigquery"
  click GCS "#cloud-storage"
  click GCSQL "#cloud-sql"
  click Kafka "#kafka"
  click Opensearch "#opensearch"
```

## Storage Comparison

Below is a list of the different storage options available to you.

| Name | Type | Availebility |  Environment | Backup |
| ---- | ---- | ------------ |  ----------- | ------ |
| [Kafka](#kafka) | Streaming | ✅ | All | No? |
| [Cloud Storage](#cloud-storage) | Object | ✅ | GCP | Yes* |
| [Cloud SQL](#cloud-sql) | Relational | ✅ | GCP | Yes |
| [BigQuery](#bigquery) | Relational | ✅ | GCP | No |
| [Elasticsearch](#elasticsearch) | Document | ⚠️ | All | No? |
| [OpenSearch](#opensearch) | Document | ✅ | All | No? |
| [InfluxDB](#influxdb) | Time Series | ⚠️ | All | No? |
| [Redis](#redis) | Key Value | ⚠️ | All | No |
| [ActiveMQ](#activemq) | Message | ⚠️ | All | No? |

> | [MongoDB](#mongodb) | Document | ⚠️ | GCP | No |

Reviewer: On-prem Mongo vil jeg tippe har backup.

Reviewer 2: Litt usikker på om MongoDB burde stå her.

Ja, vi har en guide på hvordan du kan kjøre MongoDB i clusteret, men det er ikke sikkert det er slik vi vil at de skal gjøre det (IMHO).

Alternativt så kan nogenlunde samme oppskriften brukes på de fleste databaser, så da kan vi utvide listen med MySQL, Cassandra, Oracle, Couchbase osv. med relativt enkle grep :stuck_out_tongue:.

Author: Fjerner Redis og MongoDB fra listen siden det er run it yourself.

service that is easy to use and provides a lot of flexibility. It is a good
choice for storing data that is not relational in nature.

[:octicons-arrow-right-24: Getting started with Cloud Storage](./buckets.md)

## Cloud SQL

Cloud SQL is a PostgreSQL relational database service that is provided by Google
Cloud Platform. It is a good choice for storing data that is relational in
nature.

[:octicons-arrow-right-24: Getting started with Cloud Storage](./postgres.md)

## BigQuery

BigQuery is a service that provides a relational database that is optimized for
analytical workloads. It is a good choice for storing data that is relational in
nature.

[:octicons-arrow-right-24: Getting started with Google BigQuery](./bigquery.md)

## Elasticsearch

!!! warning "Deprecated"

    Elasticsearch is deprecated and will be removed in the future. Please use
    [OpenSearch](#opensearch) instead.

Elasticsearch is a document database that is used for storing and searching
data. It is a good choice for storing data that is not relational in nature.

[:octicons-arrow-right-24: Getting started with Elasicsearch](./elastic-search.md)

> ## OpenSearch

Reviewer: Ville flyttet denne over Elasticsearch. Da blir også alt under ting som vi ikke tilbyr direkte.

Reviewer 3: Enig med Kyrre, eller bare fjerne Elastic?

Reviewer 2: Syns vi skal fjerne Elastic, det er ikke lenger mulig å få det fra Aiven og det er ingen som kjører det lenger.

Author: Fjerner Elastic siden den er deprecated.

---
description: >-
  NAIS offers several storage solutions for storing data. This page describes
  the different options and how to use them.
search:
  boost: 2
---
# Persistent Data

In this section we will discuss how to work with persistent data in your
applications and the different options available to you.

Persistent data is data that is stored on disk and survives application
restarts. This is in contrast to ephemeral data which is stored in memory
and is lost when the application is restarted.

## Responsibilities

The team is reponsible for any data that is stored in the various storage
options that are available through the platform. You can read more in the
[Data Responsibilities](./responsibilities.md) section.

> ## Avilability

Reviewer 4: ```suggestion
## Availability
```

> Some of the storage options are only aviailable from certain environments. Make

Reviewer 4: ```suggestion
Some of the storage options are only available from certain environments. Make
```

sure to check what storage options are available in your environment in the
[Storage Comparison](#storage-comparison) section below.

## What should I choose?

Sequence of questions to ask yourself when choosing the right storage option.
Choose wisely.

```mermaid
graph TD
  A[I got data!] --> B[Is it structured?]
  B --> |Yes| C[Is it events?]
  B --> |No| D[Is it files?]
  C --> |Yes| Kafka
  C --> |No| E[Is it analytical?]
  D --> |Yes| GCS[Cloud Storage]
  D --> |No| Opensearch[OpenSearch]
  E --> |Yes| GBQ[BigQuery]
  E --> |No| GCSQL[Cloud SQL]

  click GBQ "#bigquery"
  click GCS "#cloud-storage"
  click GCSQL "#cloud-sql"
  click Kafka "#kafka"
  click Opensearch "#opensearch"
```

## Storage Comparison

Below is a list of the different storage options available to you.

> | Name                            | Type        | Recommendation | Avilability | Backup |

Reviewer 4: ```suggestion
| Name                            | Type        | Recommendation | Availabilty | Backup |
```

|---------------------------------|-------------|:--------------:|:-----------:|:------:|
| [Kafka](#kafka)                 | Streaming   | ✅             | All         | Yes*   |
| [Cloud Storage](#cloud-storage) | Object      | ✅             | GCP         | Yes*   |
| [Cloud SQL](#cloud-sql)         | Relational  | ✅             | GCP         | Yes    |
| [BigQuery](#bigquery)           | Relational  | ✅             | GCP         | Yes*   |
| [Elasticsearch](#elasticsearch) | Document    | ⚠️              | All         | Yes    |
| [OpenSearch](#opensearch)       | Document    | ✅             | All         | Yes    |
| [InfluxDB](#influxdb)           | Time Series | ⚠️              | All         | Yes    |
| [IBM MQ](#ibm-mq)               | Message     | ⚠️              | All         | Yes*   |

> \* Data is highly aviailable and fault tolerent but not backed up if deleted by

Reviewer 4: ```suggestion
\* Data is highly available and fault-tolerant but not backed up if deleted by
```

# Persistent Data

In this section we will discuss how to work with persistent data in your
applications and the different options available to you.

Persistent data is data that is stored on disk and survives application
restarts. This is in contrast to ephemeral data which is stored in memory
and is lost when the application is restarted.

## Responsibilities

The team is reponsible for any data that is stored in the various storage
options that are available through the platform. You can read more in the
[Data Responsibilities](./responsibilities.md) section.

## What should I choose?

Sequence of questions to ask yourself when choosing the right storage option.
Choose wisely.

```mermaid
graph TD
  A[I got data!] --> B[Is it structured?]
  B --> |Yes| C[Is it events?]
  B --> |No| D[Is it files?]
  C --> |Yes| Kafka
  C --> |No| E[Is it analytical?]
  D --> |Yes| GCS[Cloud Storage]

>   D --> |No| Opensearch[OpenSearch]

Reviewer 2: OpenSearch kan fungere veldig bra på strukturerte data også. På grunn av hvor det kommer fra så er det lett å tenke på det som en søkemotor for "logg tekst", men det er egentlig en dokument-database med avansert søkefunksjonalitet.

Author: Ble fort veldig rotete, jeg lar denne ligge hvis noen andre vil plukke den opp senere.

```

## Storage Comparison

Below is a list of the different storage options available to you.

| Name | Type | Availebility |  Environment | Backup |
| ---- | ---- | ------------ |  ----------- | ------ |
| [Kafka](#kafka) | Streaming | ✅ | All | Yes* |
| [Cloud Storage](#cloud-storage) | Object | ✅ | GCP | Yes* |
| [Cloud SQL](#cloud-sql) | Relational | ✅ | GCP | Yes |
| [BigQuery](#bigquery) | Relational | ✅ | GCP | Yes* |
| [Elasticsearch](#elasticsearch) | Document | ⚠️ | All | Yes |
| [OpenSearch](#opensearch) | Document | ✅ | All | Yes |
| [InfluxDB](#influxdb) | Time Series | ⚠️ | All | Yes |
| [Redis](#redis) | Key Value | ⚠️ | All | No |
| [IBM MQ](#ibm-mq) | Message | ⚠️ | All | Yes* |
| [MongoDB](#mongodb) | Document | ⚠️ | GCP | No |

\* Data is highly aviailable and fault tolerent but not backed up if deleted by
mistake.

## Kafka

Kafka is a streaming platform that is used for storing and processing data. It
is a very powerful tool that can be used for a wide variety of use cases. It is
also a very complex tool that requires a lot of knowledge to use effectively.

[:octicons-arrow-right-24: Getting started with Kafka](./kafka/README.md)

> ## Cloud Storage

Reviewer 2: Legge på "(Buckets)" for å gjøre det lettere for de som ikke kan det offisielle navnet?

nature.

[:octicons-arrow-right-24: Getting started with Cloud Storage](./postgres.md)

## BigQuery

BigQuery is a service that provides a relational database that is optimized for
analytical workloads. It is a good choice for storing data that is relational in
nature.

[:octicons-arrow-right-24: Getting started with Google BigQuery](./bigquery.md)

## Elasticsearch

!!! warning "Deprecated"

    Elasticsearch is deprecated and will be removed in the future. Please use
    [OpenSearch](#opensearch) instead.

Elasticsearch is a document database that is used for storing and searching
data. It is a good choice for storing data that is not relational in nature.

[:octicons-arrow-right-24: Getting started with Elasicsearch](./elastic-search.md)

## OpenSearch

OpenSearch is a document database that is used for storing and searching data.
It is a good choice for storing data that is not relational in nature.
OpenSearch offers a drop-in replacement for Elasticsearch.

[:octicons-arrow-right-24: Getting started with OpenSearch](./open-search.md)

## InfluxDB

!!! warning "Deprecated"

    InfluxDB is deprecated and will be removed in the future. Please find a

>     suitable alternative.

Reviewer 2: For mange use cases så er BigQuery en god erstatning.
